static-check/src/types/circuits/circuit.rs

```diff
@@ -0,0 +1,177 @@
+// Copyright (C) 2019-2020 Aleo Systems Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{
+    types::circuits::{CircuitFunctionType, CircuitVariableType},
+    Attribute,
+    FunctionType,
+    SymbolTable,
+    Type,
+    TypeError,
+};
+use leo_typed::{Circuit, CircuitMember, Identifier, InputValue, Parameter, Span};
+
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+
+/// Stores circuit definition details.
+///
+/// This type should be added to the circuit symbol table for a resolved syntax tree.
+/// This is a user-defined type.
+#[derive(Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub struct CircuitType {
+    /// The name of the circuit definition.
+    pub identifier: Identifier,
+
+    /// The circuit variables.
+    pub variables: Vec<CircuitVariableType>,
+
+    /// The circuit functions.
+    pub functions: Vec<CircuitFunctionType>,
+}
+
+impl CircuitType {
+    ///
+    /// Return a new `CircuitType` from a given `Circuit` definition.
+    ///
+    /// Performs a lookup in the given symbol table if the circuit definition contains
+    /// user-defined types.
+    ///
+    pub fn new(table: &SymbolTable, unresolved: Circuit) -> Result<Self, TypeError> {
+        let circuit_identifier = unresolved.circuit_name;
+        let mut variables = vec![];
+        let mut functions = vec![];
+
+        // Resolve the type of every circuit member.
+        for member in unresolved.members {
+            match member {
+                CircuitMember::CircuitVariable(is_mutable, variable_identifier, type_) => {
+                    // Resolve the type of the circuit member variable.
+                    let type_ = Type::new_from_circuit(
+                        table,
+                        type_,
+                        circuit_identifier.clone(),
+                        circuit_identifier.span.clone(),
+                    )?;
+
+                    // Check if the circuit member variable is mutable.
+                    let attributes = if is_mutable { vec![Attribute::Mutable] } else { vec![] };
```

Comment:
if I'm not mistaken, since there are just 2 possible `Attribute`s, using a `Vec` here seems wasteful; I'd instead use a `SmallVec` or, even cheaper, some `enum` that is either `Mutable`, `Static` **or** a combination of both